File: src/demobot.py
import time

from lib.libwallaby import libwallaby
from lib.wheels import *

red_channel = 0
green_channel = 1

arm_servo = Servo(port=0, delay=20)

spinner_motor = Motor(port=2, speed=1)

wheels = Wheels(
    left_motor=Motor(port=0, speed=1),
    right_motor=Motor(port=1, speed=1),
    left_offset=0.91,  # if <1, veers to the right
    right_offset=1.0,  # if <1, veers to the left
)


half_turn = 150  # instead of 180 to counter offset


class Color(Enum):
    red = 0
    green = 1


def main():
    libwallaby.camera_close()
    libwallaby.camera_open()

    raise_arm()

    # Collect only the red poms

    number_of_poms = 8
    number_of_red_poms = 0
    distance_between_poms = inches(5.25)

    target_color = Color.red

    for _ in range(number_of_poms - 1):  # the first pom is skipped
        wheels.turn(TurnDirection.left, 86)

        def collect():
            nonlocal number_of_red_poms

            wheels.start(Direction.forward)
            run_until_poms(None, timeout=3)
            wheels.stop()

            align_wheels_to_pom()

            wheels.drive(Direction.forward, cm(1))

            color = detect_color()
            print("COLOR:", color)

            if color == target_color:
                lower_arm()
                collect_pom()
                raise_arm()
                number_of_red_poms += 1

        with_reset_wheels(collect)

        wheels.turn(TurnDirection.right, 80)

        wheels.drive(Direction.forward, distance_between_poms)

    # Wait for Create

    # TODO

    # Drive to the material transport

    wheels.drive(Direction.forward, m(0.5))

    # Dispense the red poms

    dispense_poms_v2()

    # Turn around

    wheels.turn(TurnDirection.left, 75)
    wheels.drive(Direction.forward, cm(13))
    wheels.turn(TurnDirection.left, 82)
    wheels.drive(Direction.forward, m(0.45))

    # Collect the remaining (green) poms

    number_of_green_poms = number_of_poms - number_of_red_poms

    for _ in range(number_of_green_poms):
        wheels.start(Direction.forward)
        run_until_poms(None, timeout=5)
        wheels.stop()

        align_wheels_to_pom()

        wheels.drive(Direction.forward, cm(2))

        lower_arm()
        collect_pom()
        raise_arm()

    # Turn around and drive to the material transport

    wheels.turn(TurnDirection.right, half_turn)
    wheels.drive(Direction.forward, m(1.3))

    # Dispense the green poms

    wheels.turn(TurnDirection.left, 30)
    dispense_poms_v2()

    # TODO


def align_wheels_to_pom():
    color = detect_color()
    if color is None:
        return
    channel = channel_of(color)

    if libwallaby.get_object_count(channel) == 0:
        return

    center_x = libwallaby.get_camera_width() / 2
    threshold = 5

    wheels.left_motor.speed = 0.1
    wheels.right_motor.speed = 0.1

    while True:
        libwallaby.camera_update()

        diff = libwallaby.get_object_center_x(channel, 0) - center_x

        if diff < -threshold:
            wheels.start_turn(TurnDirection.left)
        elif diff > threshold:
            wheels.start_turn(TurnDirection.right)
        else:
            wheels.stop()
            wheels.left_motor.speed = 1
            wheels.right_motor.speed = 1
            break


def dispense_poms_v2():
    raise_arm_halfway()
    spinner_motor.start(Direction.reverse)

    for _ in range(5):
        shake()

    spinner_motor.stop()
    raise_arm()


def raise_arm():
    arm_servo.set(420)


def raise_arm_halfway():
    arm_servo.set(620)


def lower_arm():
    arm_servo.set(680)


def collect_group():
    colors = []

    lower_arm()

    color = with_reset_wheels(collect_pom)
    if color:
        colors.append(color)

    wheels.turn(TurnDirection.right, 90)
    wheels.drive(Direction.forward, cm(10))
    wheels.turn(TurnDirection.left, 90)

    def collect_2():
        for _ in range(2):
            color = collect_pom()
            if color:
                colors.append(color)

    with_reset_wheels(collect_2)


def detect_pom():
    wheels.start(Direction.forward)
    while not pom_detected():
        pass
    wheels.stop()

    return detect_color()


def collect_pom():
    spinner_motor.start(Direction.forward)
    wheels.start(Direction.forward)
    libwallaby.msleep(950)
    spinner_motor.stop()
    wheels.stop()


def detect_color():
    for _ in range(10):
        libwallaby.camera_update()

    if libwallaby.get_object_count(red_channel) > 0:
        return Color.red
    elif libwallaby.get_object_count(green_channel) > 0:
        return Color.green
    else:
        return None


def shake():
    shake_angle = 5

    for _ in range(5):
        wheels.turn(TurnDirection.left, shake_angle)
        wheels.turn(TurnDirection.right, shake_angle)


def channel_of(color):
    if color == Color.red:
        return red_channel
    elif color == Color.green:
        return green_channel
    else:
        raise ValueError("expected color")


def pom_detected():
    libwallaby.camera_update()

    return (
        libwallaby.get_object_count(red_channel) > 0
        or libwallaby.get_object_count(green_channel) > 0
    )


def run_until_poms(f, timeout):
    start = time.time()
    while time.time() - start <= timeout:
        if f:
            f()

        if pom_detected():
            break


def run_until_no_poms(f, timeout):
    start = time.time()
    while time.time() - start <= timeout:
        if f:
            f()

        if not pom_detected():
            break


def with_reset_wheels(f, direction=Direction.reverse, padding=0):
    value, distance = get_wheel_distance_after(f)
    wheels.drive(direction, distance - padding)
    return value


def get_wheel_distance_after(f):
    initial_position_left = libwallaby.get_motor_position_counter(
        wheels.left_motor.port
    )
    initial_position_right = libwallaby.get_motor_position_counter(
        wheels.right_motor.port
    )

    value = f()

    final_position_left = libwallaby.get_motor_position_counter(wheels.left_motor.port)
    final_position_right = libwallaby.get_motor_position_counter(
        wheels.right_motor.port
    )

    left_distance = final_position_left - initial_position_left
    right_distance = final_position_right - initial_position_right

    return value, (
        (left_distance if left_distance < right_distance else right_distance)
        / motor_pwm_ticks
        / motor_travel_time_1_cm
    )

File: src/demobot.c
#if __has_include(<kipr/wombat.h>)
#include <kipr/wombat.h>
#else
#error "Please compile on the Wombat"
#endif

#include <stdio.h>
#include <stdlib.h>
#include "wheels.h"

void raise_arm();
void lower_arm();
void reset_splitter();
void split_pom_left();
void split_pom_right();
void drive_until_reached_pom();

const int red_channel = 0;
const int green_channel = 1;

const Servo arm_servo = {
    .port = 0,
};

const Servo splitter_servo = {
    .port = 1,
};

const Motor spinner_motor = {
    .port = 2,
    .speed = 1.0,
};

const Wheels wheels = {
    .left_motor = {
        .port = 0,
        .speed = 1.0,
    },
    .right_motor = {
        .port = 1,
        .speed = 1.0,
    },
    .left_offset = 0.9975, // if <1, veers to the right
    .right_offset = 1.0, // if <1, veers to the left
};

/*

lift arm

repeat 8 times:
    if green:
        move servo left
    else:
        move servo right
    move forward

lower arm

THEN: go back around to collect poms

*/

int main() {
    camera_open();
    raise_arm();
    reset_splitter();

    int pom_width = CM(2);
    int space_between_poms = IN(6) - pom_width;

    int initial_position_left = get_motor_position_counter(wheels.left_motor.port);
    int initial_position_right = get_motor_position_counter(wheels.right_motor.port);

    for (int i = 0; i < 5; i++) {
        drive_until_reached_pom();

        int pom_is_red = get_object_count(red_channel) > 0;
        int advance_distance = CM(10);

        if (pom_is_red) {
            printf("Red!\n"); fflush(stdout);
            split_pom_right();
            drive_wheels(wheels, FORWARD, advance_distance);
            split_pom_left();
        } else {
            printf("Green!\n"); fflush(stdout);
            split_pom_left();
            drive_wheels(wheels, FORWARD, advance_distance);
            split_pom_right();
        }

        reset_splitter();
    }

    msleep(2000); // FIXME: TEMPORARY

    // Go back and collect the poms
    // TODO: Put into function

    int final_position_left = get_motor_position_counter(wheels.left_motor.port);
    int final_position_right = get_motor_position_counter(wheels.right_motor.port);

    int left_distance = final_position_left - initial_position_left;
    int right_distance = final_position_right - initial_position_right;

    int distance = (left_distance < right_distance ? left_distance : right_distance) / MOTOR_PWM_TICKS / MOTOR_TRAVEL_TIME_1_CM;

    drive_wheels(wheels, REVERSE, distance - CM(15));

    turn_wheels(wheels, LEFT, 120);
    drive_wheels(wheels, FORWARD, CM(18));
    turn_wheels(wheels, RIGHT, 120);

    lower_arm();

    motor(spinner_motor.port, 100);
    drive_wheels(wheels, FORWARD, space_between_poms * 3);
    off(spinner_motor.port);

    camera_close();

    return 0;
}

void raise_arm() {
    set_servo(arm_servo, 1040);
}

void lower_arm() {
    set_servo(arm_servo, 1250);
}

void reset_splitter() {
    set_servo(splitter_servo, 1280);
}

void split_pom_left() {
    set_servo(splitter_servo, 0);
}

void split_pom_right() {
    set_servo(splitter_servo, 2047);
}

void drive_until_reached_pom() {
    for (int i = 0; i < 10; i++) {
        camera_update();
    }

    int left_velocity = calculate_velocity(wheels.left_motor.speed, FORWARD, wheels.left_offset);
    int right_velocity = calculate_velocity(wheels.right_motor.speed, FORWARD, wheels.right_offset);

    move_at_velocity(wheels.left_motor.port, left_velocity);
    move_at_velocity(wheels.right_motor.port, right_velocity);

    while (1) {
        camera_update();

        if (get_object_count(red_channel) > 0 || get_object_count(green_channel) > 0) {
            break;
        }
    }

    force_stop_wheels(wheels);
}
